cpp/tests/parallel_write_test.cc

```diff
@@ -0,0 +1,122 @@
+            ms.addColumn(table_desc, storage_manager);
+        }
+
+        return std::make_shared<casacore::TableProxy>(ms);
+      };
+
+      ASSERT_OK_AND_ASSIGN(table_proxy_, arcae::SafeTableProxy::Make(factory));
+    }
+};
+
+// Parallel SafeTableProxy writes
+TEST_F(WriteTests, Parallel) {
+  ASSERT_TRUE(table_proxy_);
+  ASSERT_EQ(table_proxy_->nRow(), knrow);
+
+  auto data_result = table_proxy_->run(
+    [](const casacore::TableProxy & proxy) -> arrow::Result<casacore::Array<casacore::Complex>> {
+        auto table = proxy.table();
+        auto table_column = casacore::TableColumn(table, "MODEL_DATA");
+        auto column = casacore::ArrayColumn<casacore::Complex>(table_column);
+        return column.getColumn();
+    });
+
+  ASSERT_OK_AND_ASSIGN(auto data, data_result);
+
+  for(auto [i, it] = std::tuple{0, data.begin()}; it != data.end(); ++it, ++i) {
+    *it = i;
+  }
+
+  auto name_result = table_proxy_->run(
+    [data = std::move(data)](const casacore::TableProxy & proxy) -> arrow::Result<std::string> {
+        auto table = proxy.table();
+        auto table_column = casacore::TableColumn(table, "MODEL_DATA");
+        auto column = casacore::ArrayColumn<casacore::Complex>(table_column);
+        column.putColumn(data);
+        return table.tableName();
+    });
+
+  ASSERT_OK_AND_ASSIGN(auto name, name_result);
+  table_proxy_.reset();
+
+  std::vector<std::shared_ptr<arcae::SafeTableProxy>> writers;
+
+  for(std::size_t i=0; i < knthreads; ++i) {
+    ASSERT_OK_AND_ASSIGN(auto writer, arcae::OpenTable(name, false));
+    writers.emplace_back(std::move(writer));
+  }
+
+  ASSERT_OK_AND_ASSIGN(auto pool, arrow::internal::ThreadPool::Make(knthreads));
+
+  static constexpr std::size_t inc = 2;
+
+  std::vector<arrow::Future<bool>> futures;
+
+  for(auto [r, c] = std::tuple{std::size_t{0}, std::size_t{0}}; r < knrow; r += inc, ++c) {
+    std::size_t end = std::min(r + inc, knrow);
+    auto result = arrow::DeferNotOk(pool->Submit(
+        [start=r, nrow=end-r, writer=writers[c % knthreads]]() mutable {
+            return writer->run([start=start, nrow=nrow](casacore::TableProxy & proxy) mutable -> arrow::Result<bool> {
+                auto table = proxy.table();
+                auto table_column = casacore::TableColumn(table, "MODEL_DATA");
+                auto & column_desc = table_column.columnDesc();
+                auto column = casacore::ArrayColumn<casacore::Complex>(table_column);
+                column.setMaximumCacheSize(1);
+                auto data = casacore::Array<casacore::Complex>(casacore::IPosition({kncorr, knchan, int(nrow)}));
+                data.set(start);
+
+                try {
+                    column.putColumnRange(casacore::Slice(start, nrow), data);
+                    table.flush();
+                } catch(std::exception & e) {
+                    return arrow::Status::Invalid("Write failed ", e.what());
+                }
```

Comment:
I believe this creates 2 rows worth of data, writes it out to the column on disk and performs a flush (maybe this should be a column flush?)